### PR TITLE
Task 3

### PR DESCRIPTION
**Pull request by Pokemon3108**

### CODE REVIEW

File: Task2/persistence/src/main/java/com/epam/esm/dao/impl/UserDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+        CriteriaBuilder criteriaBuilder = em.getCriteriaBuilder();
+        CriteriaQuery<User> criteriaQuery = criteriaBuilder.createQuery(User.class);
+        Root<User> userRoot = criteriaQuery.from(User.class);
+
+        Join<User, Order> join = userRoot.join("orders", JoinType.LEFT);
+        criteriaQuery.select(userRoot)
+                .groupBy(userRoot.get("id"))
+                .orderBy(criteriaBuilder.desc(criteriaBuilder.sum(join.get("cost"))));
+        TypedQuery<User> query = em.createQuery(criteriaQuery);
```
Comment (evgeniyyaskevich) [inline Task2/persistence/src/main/java/com/epam/esm/dao/impl/UserDaoImpl.java:59]: What about setting maxResult = 1 to query and avoid querying of a big amount of users?

File: Task2/persistence/src/main/java/com/epam/esm/dto/converter/GiftCertificateDtoConverter.java [hunk 2]
```diff
@@ -14,15 +16,25 @@
 @Component
 public class GiftCertificateDtoConverter {
```
Comment (evgeniyyaskevich) [inline Task2/persistence/src/main/java/com/epam/esm/dto/converter/GiftCertificateDtoConverter.java:19]: Is it reasonable to extend the abstract Converter class here too?

File: Task2/service/src/main/java/com/epam/esm/impl/GiftCertificateServiceImpl.java [hunk 3]
```diff
@@ -139,39 +131,35 @@ public void update(GiftCertificateDto certificateDto) {
...
 
-        List<GiftCertificateDto> certificateDtos = new ArrayList<>();
-        for (GiftCertificate c : certificatesWithParams) {
-            Set<Tag> tags = new HashSet<>();
-            Set<Long> tagsIds = tagDao.readCertificateTagsIdsByCertificateId(c.getId());
-            if (!tagsIds.isEmpty()) {
-                tags = tagDao.readTagsByIds(tagsIds);
-            }
-            certificateDtos.add(dtoConverter.convertToDto(c, tags));
+        final int maxSize = 100;
```
Comment (evgeniyyaskevich) [inline Task2/service/src/main/java/com/epam/esm/impl/GiftCertificateServiceImpl.java:149]: It is better to move it to static final class variable.

File: Task2/service/src/main/java/com/epam/esm/impl/GiftCertificateServiceImpl.java [hunk 4]
```diff
@@ -139,39 +131,35 @@ public void update(GiftCertificateDto certificateDto) {
...
+        if (size > maxSize || size < 0) {
+            throw new MaxSizeLimitException(maxSize);
+        }
+        if (page < 0 || certificateDao.getAllCount() < (page - 1) * size) {
+            throw new NoPageException(page, size);
         }
+        List<GiftCertificateDto> certificatesWithParams = certificateDao
+                .findCertificateByParams(page, size, certificateDto);
+        return new ArrayList<>(certificatesWithParams);
```
Comment (evgeniyyaskevich) [inline Task2/service/src/main/java/com/epam/esm/impl/GiftCertificateServiceImpl.java:158]: Why do we need new list here?

File: Task2/persistence/src/main/java/com/epam/esm/dao/impl/TagDaoImpl.java [hunk 5]
```diff
@@ -1,189 +1,129 @@
...
-    @Override
-    public Set<Tag> readTagsByIds(Set<Long> ids) {
-
-        String inSql = String.join(",", Collections.nCopies(ids.size(), "?"));
-
-        List<Tag> tags = jdbcTemplate.query(String.format(READ_TAGS_BY_TAGS_IDS, inSql),
-                tagMapper, ids.toArray());
+    public Optional<TagDto> readTheMostPopularTagOfRichestUser() {
+
+        Query query = em.createNativeQuery(
```
[the remark this answers is not in the document]
Comment (Pokemon3108) [reply Task2/persistence/src/main/java/com/epam/esm/dao/impl/TagDaoImpl.java:113]: createNativeQuery doen't work with TypedQuery

**Pull request merged**